### PR TITLE
Add leaderboard stats, periodic updates via fastapi-utils

### PR DESCRIPTION
**Pull request by andreaskoepf**
Provide new leaderboard REST endpoints: 
`/api/v1/leaderboards/{day,week,month,total}`

Example response:

```
{
  "time_frame": "day",
  "leaderboard": [
    {
      "user_rank": 0,
      "user_id": "38312e49-8cd6-4e7c-8bad-c626e452b1e7",
      "username": "__dummy_user__",
      "auth_method": "local",
      "display_name": "Dummy User",
      "leader_score": 2740,
      "base_date": "2023-01-14T00:42:14.937437",
      "modified_date": "2023-01-15T00:42:14.937534",
      "prompts": 74,
      "replies_assistant": 261,
      "replies_prompter": 109,
      "labels_simple": 0,
      "labels_full": 0,
      "rankings_total": 0,
      "rankings_good": 0,
      "accepted_prompts": 74,
      "accepted_replies_assistant": 261,
      "accepted_replies_prompter": 109,
      "reply_ranked_1": 22,
      "reply_ranked_2": 22,
      "reply_ranked_3": 22,
      "streak_last_day_date": null,
      "streak_days": null
    }
}
```

### CODE REVIEW

File: backend/oasst_backend/api/v1/leaderboards.py [hunk 1]
```diff
@@ -1,26 +1,50 @@
...
+from oasst_backend.user_stats_repository import UserStatsRepository, UserStatsTimeFrame
 from oasst_shared.schemas.protocol import LeaderboardStats
 from sqlmodel import Session
 
 router = APIRouter()
 
 
-@router.get("/create/assistant")
-def get_assistant_leaderboard(
+@router.get("/day")
```
Comment (yk) [inline backend/oasst_backend/api/v1/leaderboards.py:13]: all these routes could probably be unified to save some code

File: backend/oasst_backend/models/user_stats.py [hunk 2]
```diff
@@ -40,14 +41,27 @@ class UserStats(SQLModel, table=True):
...
+    reply_ranked_3: int = 0
 
     # only used for time span "total"
     streak_last_day_date: Optional[datetime] = Field(nullable=True)
     streak_days: Optional[int] = Field(nullable=True)
+
+    def compute_leader_score(self) -> int:
+        return (
+            self.prompts
+            + self.replies_assistant * 4
```
Comment (yk) [inline backend/oasst_backend/models/user_stats.py:55]: how about making these constants into settings?

File: oasst-shared/oasst_shared/utils.py [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+from loguru import logger
+
 
 def utcnow() -> datetime:
     """Return the current utc date and time with tzinfo set to UTC."""
     return datetime.now(timezone.utc)
+
+
+def log_timing(func):
+    def wrapped(*args, **kwargs):
```
Comment (yk) [inline oasst-shared/oasst_shared/utils.py:13]: just for prettiness, could use `functools.wraps`

File: backend/alembic/versions/2023_01_15_0002-7c98102efbca_change_user_stats_ranking_counts.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        sa.ForeignKeyConstraint(
+            ["user_id"],
+            ["user.id"],
+        ),
+        sa.PrimaryKeyConstraint("user_id", "time_frame"),
+    )
+    # ### end Alembic commands ###
+
+
+def downgrade() -> None:
```
Comment (yk) [inline backend/alembic/versions/2023_01_15_0002-7c98102efbca_change_user_stats_ranking_counts.py:54]: the upgrade drops and recreates the user_stats table, while the downgrade just modifies the columns. it's probably fine though

**Review by yk — APPROVED** [review]
awesome

File: backend/alembic/versions/2023_01_15_0002-7c98102efbca_change_user_stats_ranking_counts.py (hunk 4 above)
Comment (andreaskoepf) [reply backend/alembic/versions/2023_01_15_0002-7c98102efbca_change_user_stats_ranking_counts.py:54]: Yes, intentionally. Somehow alembic didn't manage it to add a second column to the PR as a change, e.g. this table has  a combined user_id + time_frame PK and before had only user_id. The table wasn't used before...

File: backend/main.py [hunk 5]
```diff
@@ -195,6 +197,17 @@ def ensure_tree_states():
         logger.exception("TreeManager.ensure_tree_states() failed.")
 
 
+@app.on_event("startup")
+@repeat_every(seconds=60 * 60, wait_first=False)  # 1 hour
+def remove_expired_tokens_task() -> None:
```
Comment (andreaskoepf) [inline backend/main.py:202]: ok, I have to change the name here

File: backend/oasst_backend/api/v1/leaderboards.py (hunk 1 above)
Comment (andreaskoepf) [reply backend/oasst_backend/api/v1/leaderboards.py:13]: ok

File: backend/oasst_backend/models/user_stats.py (hunk 2 above)
Comment (andreaskoepf) [reply backend/oasst_backend/models/user_stats.py:55]: We need to find a "stable" formula for the leader-score .. calculation is "magic" right now .. the exact way how to calculate it needs more discussion, e.g. could be based on acceptance rations or include negative values..

File: oasst-shared/oasst_shared/utils.py (hunk 3 above)
Comment (andreaskoepf) [reply oasst-shared/oasst_shared/utils.py:13]: I made it now a param + non-param decorator .. with optional logging of the kwargs. regarding `functool.wraps` ..that makes sure that docstring and function name are preserved .. right?

Comment (yk) [reply oasst-shared/oasst_shared/utils.py]: yes, it's just for clarity

**Pull request merged**